pkg/s3/client.go

```diff
@@ -0,0 +1,101 @@
+package s3
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	sync "github.com/bacalhau-project/golang-mutex-tracer"
+)
+
+type ClientWrapper struct {
+	S3         *s3.Client
+	Downloader *manager.Downloader
+	Uploader   *manager.Uploader
+	Endpoint   string
+	Region     string
+}
+
+type ClientProviderParams struct {
+	AWSConfig aws.Config
+}
+
+type ClientProvider struct {
+	awsConfig aws.Config
+	clients   map[string]*ClientWrapper
+	mu        sync.RWMutex
```

Comment:
readability opinion: this mutex is guarding `clients`, give it a hat and call it `clientsMu` or reorder this structure as:
```go
type ClientProvider struct {
	awsConfig aws.Config
	
	mu        sync.RWMutex
	clients   map[string]*ClientWrapper
}
```
to indicate the fields the mutex is guarding.